demo/src/main/java/com/smashingboxes/surelockdemo/LoginActivity.kt

```diff
@@ -0,0 +1,346 @@
+                } catch (e: SurelockInvalidKeyException) {
+                    surelockStorage.clearAll()
+                    showFingerprintLoginInvalidated()
+                }
+
+                fingerprint_checkbox.visibility = View.GONE
+            }
+        }
+    }
+
+    override fun onStop() {
+        //Clear storage so users don't have to clear app data to try out new credentials
+        surelockStorage.clearAll()
+        super.onStop()
+    }
+
+    private fun getSurelock(): Surelock {
+        if (surelock == null) {
+            surelock = Surelock.Builder(this)
+                .withDefaultDialog(R.style.SurelockDemoDialog)
+                .withKeystoreAlias(KEYSTORE_KEY_ALIAS)
+                .withFragmentManager(fragmentManager)
+                .withSurelockFragmentTag(FINGERPRINT_DIALOG_FRAGMENT_TAG)
+                .withSurelockStorage(surelockStorage)
+                .build()
+        }
+        return surelock!!
+    }
+
+    private fun showFingerprintLoginInvalidated() {
+        val builder = AlertDialog.Builder(this)
+        builder.setTitle(R.string.sl_login_error)
+            .setMessage(R.string.sl_re_enroll)
+            .setPositiveButton(android.R.string.ok) { dialog, _ -> dialog.dismiss() }.show()
+    }
+
+    private fun userWantsToUseFingerprintToLogin(): Boolean {
+        return preferenceHelper.contains(KEY_CRE_DEN_TIALS)
+    }
+
+    private fun validateLoginForm(usernameInput: String?, passwordInput: String?): Boolean {
+        var errorsExist = false
+        var focusView: View? = null
+
+        if (TextUtils.isEmpty(usernameInput)) {
+            username_input_layout.error = getString(R.string.error_field_required)
+            focusView = username
+            errorsExist = true
+        } else {
+            username_input_layout.error = null
+        }
+
+        if (TextUtils.isEmpty(passwordInput)) {
+            password_input_layout.error = getString(R.string.error_field_required)
+            focusView = password
+            errorsExist = true
+        } else {
+            password_input_layout.error = null
+        }
+
+        if (errorsExist) {
+            focusView!!.requestFocus()
+        }
+        return errorsExist
+    }
+
+    private fun handleLoginClick() {
+        val username = username.text.toString()
+        val password = password.text.toString()
+
+        if (fingerprint_checkbox.isChecked && Surelock.fingerprintAuthIsSetUp(this, true)) {
+            attemptLoginForFingerprintEnrollment(username, password)
+        } else {
+            attemptLogin(username, password)
+        }
+    }
+
+    private fun attemptLogin(usernameInput: String?, passwordInput: String?) {
+        // Don't allow multiple login attempts in a row
+        if (authTask != null) {
+            return
+        }
+
+        if (!validateLoginForm(usernameInput, passwordInput)) {
+            showProgress(true)
+
+            authTask = UserLoginTask(false)
```

Comment:
When making constructor/function calls where there is a mysterious Boolean parameter, you should try to always set the name of the parameter explicitly, like so:
`UserLoginTask(withFingerprintEnrollment = false)`
Note: there's at least one other place where you're missing this.